crates/node-api/src/evm/traits.rs

```diff
@@ -0,0 +1,133 @@
+use std::time::Duration;
+
+use reth_primitives::{
+    revm::env::fill_block_env, Address, BlockWithSenders, ChainSpec, Header, Receipt, Transaction,
+    U256,
+};
+use revm::Database;
+use revm_primitives::{BlockEnv, CfgEnvWithHandlerCfg, SpecId, TxEnv};
+
+use super::BundleStateWithReceipts;
+
+/// Trait for configuring the EVM.
+pub trait EvmConfig: ConfigureEvmEnv {
+    /// The type that can executes transactions and full blocks.
+    type Executor: BlockExecutor;
+
+    /// A hook that allows to modify the EVM before execution.
+    fn evm(&self, db: impl Database) -> Self::Executor;
+
+    /// A hook that allows to modify the EVM before execution, with an inspector.
+    fn evm_with_inspector<I>(&self, db: impl Database, inspector: I) -> Self::Executor;
+}
+
+/// This represents the set of methods used to configure the EVM before execution.
+pub trait ConfigureEvmEnv: Send + Sync + Unpin + Clone {
+    /// The type of the transaction metadata.
+    type TxMeta;
+
+    /// Fill transaction environment from a [Transaction] and the given sender address.
+    fn fill_tx_env<T>(tx_env: &mut TxEnv, transaction: T, sender: Address, meta: Self::TxMeta)
+    where
+        T: AsRef<Transaction>;
+
+    /// Fill [CfgEnvWithHandlerCfg] fields according to the chain spec and given header
+    fn fill_cfg_env(
+        cfg_env: &mut CfgEnvWithHandlerCfg,
+        chain_spec: &ChainSpec,
+        header: &Header,
+        total_difficulty: U256,
+    );
+
+    /// Convenience function to call both [fill_cfg_env](ConfigureEvmEnv::fill_cfg_env) and
+    /// [fill_block_env].
+    fn fill_cfg_and_block_env(
+        cfg: &mut CfgEnvWithHandlerCfg,
+        block_env: &mut BlockEnv,
+        chain_spec: &ChainSpec,
+        header: &Header,
+        total_difficulty: U256,
+    ) {
+        Self::fill_cfg_env(cfg, chain_spec, header, total_difficulty);
+        let after_merge = cfg.handler_cfg.spec_id >= SpecId::MERGE;
+        fill_block_env(block_env, chain_spec, header, after_merge);
+    }
+}
+
+/// An executor capable of executing a block.
+pub trait BlockExecutor {
```

Comment:
do we dedupe with https://github.com/paradigmxyz/reth/blob/2a5efb2a99ca50a45e41c50b5ef0e6d838d46b28/crates/storage/provider/src/traits/executor.rs#L21 eventually?

Reply:
yep, actually I think I'll move this to a new crate that `reth-revm` and `node-api` will both depend on, instead of including this (and relevant types) in `node-api`. Mainly to avoid circular dependencies, and avoid making `node-api` gigantic

Reply:
actually, hmm, need to think more about this, this might not work